### PR TITLE
Make storage detachable/attachable

### PR DESCRIPTION
**Pull request by axw**
## Description of change

When detaching storage from a unit, we now
leave the storage in the model, and just
clear the storage instance's ownership. The
storage can then be attached to another unit
using the new State.AttachStorage method.
Hook up the attach-storage backend so it calls
State.AttachStorage.

The existing model storage cleanups have been
replaced with one that destroys storage
instances, rather than volumes and filesystems.
The machine storage will be destroyed as a
consequence.

## QA steps

1. juju bootstrap aws
2. juju deploy storagetest -n2
3. juju add-storage storagetest/0 disks=ebs
(wait)
4. juju detach-storage disks/0
5. juju attach-storage storagetest/1 disks/0
(wait - should fail because storagetest/0 and storagetest/1 are in different AZs)
6. juju detach-storage disks/0
7. juju add-machine zone=(same zone as storagetest/0's machine)
8. juju add-unit storagetest --to (machine ID from step 7)
9. juju attach-storage storagetest/2 disks/0
(wait - should attach cleanly)
10. juju destroy-controller -y aws-...

## Documentation changes

When this branch lands into develop, after 2.2 is released, we need to document detach-storage and attach-storage in the devel docs.

## Bug reference

https://bugs.launchpad.net/bugs/1592887

### CODE REVIEW

File: state/addmachine.go [hunk 1]
```diff
@@ -597,12 +602,14 @@ func (st *State) machineStorageOps(
 		}
 		volumeOps = append(volumeOps, ops...)
 		volumeAttachments = append(volumeAttachments, volumeAttachmentTemplate{
-			tag, v.Attachment,
+			tag, v.Attachment, false,
+		})
+	}
+	for tag, volumeAttachment := range args.volumeAttachments {
+		volumeAttachments = append(volumeAttachments, volumeAttachmentTemplate{
+			tag, volumeAttachment, true,
```
Comment (wallyworld) [inline state/addmachine.go]: would be nice to have a const ExistingStorage=true and NewStorage=false or something and use the consts instead of true/false

**Comment by wallyworld** [comment]
Also, is it worth adding a feature test or two?

File: state/addmachine.go (hunk 1 above)
Comment (axw) [inline state/addmachine.go]: I've added function local consts to make it clearer.

File: apiserver/common/storagecommon/storage.go [hunk 2]
```diff
@@ -57,6 +58,10 @@ type StorageInterface interface {
 	// BlockDevices returns information about block devices published
 	// for the specified machine.
 	BlockDevices(names.MachineTag) ([]state.BlockDeviceInfo, error)
+
+	// UnitStorageAttachments returns the storage attachments for the
+	// identified unit.
```
Comment (axw) [inline apiserver/common/storagecommon/storage.go]: sure, though several of the methods above use identified as well :)  I've changed them all

File: apiserver/application/application.go [hunk 3]
```diff
@@ -683,11 +684,16 @@ func (api *API) DestroyUnit(args params.Entities) (params.DestroyUnitResults, er
 			return nil, errors.Errorf("unit %q is a subordinate", name)
 		}
 		var info params.DestroyUnitInfo
-		storage, err := common.UnitStorage(api.backend, unit.UnitTag())
+		storage, err := storagecommon.UnitStorage(api.backend, unit.UnitTag())
+		if err != nil {
+			return nil, err
```
Comment (axw) [inline apiserver/application/application.go]: done

**Comment by axw** [comment]
$$merge$$

**Pull request merged**